From 5cc29564fe8a757866130366c6109f7c42fe7638 Mon Sep 17 00:00:00 2001
From: Matthew Vogt <matthew.vogt@jollamobile.com>
Date: Tue, 8 Jan 2013 15:42:58 +1100
Subject: [PATCH 08/18] Improve QSystemDeviceInfo::uniqueDeviceID value

For non-specific linux variants, if DBUS is present prefer to return
the exported localMachineId value.  Otherwise fallback to a random but
persistent UUID value, which is more likely to be unique than the result
of gethostid().

Change-Id: I728fd24087caae3ff95693c224887bbcc03e7484
Signed-off-by: Matthew Vogt <matthew.vogt@jollamobile.com>
---
 src/systeminfo/linux/qsysteminfo_linux_common.cpp |   31 +++++++++++++--------
 1 file changed, 20 insertions(+), 11 deletions(-)

diff --git a/src/systeminfo/linux/qsysteminfo_linux_common.cpp b/src/systeminfo/linux/qsysteminfo_linux_common.cpp
index e8aa6f8..c36b1ca 100644
--- a/src/systeminfo/linux/qsysteminfo_linux_common.cpp
+++ b/src/systeminfo/linux/qsysteminfo_linux_common.cpp
@@ -2821,6 +2821,7 @@ QByteArray QSystemDeviceInfoLinuxCommonPrivate::uniqueDeviceID()
     QCryptographicHash hash(QCryptographicHash::Sha1);
 #if !defined(QT_NO_DBUS)
 #if !defined(QT_NO_HAL)
+    // For backward compatibility, return HAL UUID if available
     if (halIsAvailable) {
         QHalDeviceInterface iface("/org/freedesktop/Hal/devices/computer", this);
         QString id;
@@ -2831,19 +2832,27 @@ QByteArray QSystemDeviceInfoLinuxCommonPrivate::uniqueDeviceID()
         }
     }
 #endif // QT_NO_HAL
-#if defined(Q_WS_MEEGO)
-    QDBusInterface connectionInterface("org.freedesktop.PolicyKit1", "/org/freedesktop/PolicyKit1/Authority", "org.freedesktop.DBus.Peer",
-                                       QDBusConnection::systemBus(), this);
-    if (!connectionInterface.isValid())
-        qDebug() <<connectionInterface.lastError().message()<< "not valid";
-
-    QDBusReply<QString> reply = connectionInterface.call("GetMachineId");
-    QString uid = reply.value();
-    hash.addData(uid.toLocal8Bit());
+
+    // Return the DBUS machine ID
+    QByteArray dbusId = QDBusConnection::localMachineId();
+    hash.addData(dbusId);
     return hash.result().toHex();
-#endif // Q_WS_MEEGO
 #endif // QT_NO_DBUS
-    hash.addData(QString::number(gethostid()).toLocal8Bit());
+
+    QSettings systemInfoSettings(QLatin1String("Nokia"), QLatin1String("SystemInfo"));
+
+    const QString key(QLatin1String("localMachineId"));
+    QVariant value = systemInfoSettings.value(key);
+    if (!value.isValid()) {
+        // Create a persistent value for this purpose
+        QByteArray printable = QUuid::createUuid().toString().toLocal8Bit();
+        systemInfoSettings.setValue(key, QVariant(printable));
+
+        hash.addData(printable);
+    } else {
+        hash.addData(value.value<QByteArray>());
+    }
+
     return hash.result().toHex();
 }
 
-- 
1.7.10.3

